msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java

```diff
@@ -1914,4 +2040,225 @@ else if (toolId.equals(DiscussionForumService.MESSAGES_TOOL_ID))
+		  }
+	  }
+		
+	  return rrepMsg;
+  }
+  
+  public Map<User, Boolean> getRecipients(List recipients) {     
+	  Map<User, Boolean> returnSet = new HashMap<>();
+    
+	  /** get List of unfiltered course members */
+	  for (Iterator iterator = recipients.iterator(); iterator.hasNext();) {
+		  PrivateMessageRecipient element = (PrivateMessageRecipient) iterator.next();
+		  try {
+			  returnSet.put(userDirectoryService.getUser(element.getUserId()), false);
+		  } catch (UserNotDefinedException e) {
+			  log.warn("PrivateMessageManagerImpl.getRecipients: exception: " + e.getMessage(), e);
+		  }
+	  }
+	  //now add them all back
+	  return returnSet;
+  }
+
+  private boolean errRecipient(List recipients, PrivateMessage rrepMsg) {
+	  for (Iterator iterator = recipients.iterator(); iterator.hasNext();) {
+		  PrivateMessageRecipient element = (PrivateMessageRecipient) iterator.next();
+		  if (element.getUserId().equals(rrepMsg.getCreatedBy())) {
+			  return false;
+		  }
+	  }
+	  return true;
+  }
+
+  private boolean errMaxMessageRespond(PrivateMessage rrepMsg) {
+	  int max = ServerConfigurationService.getInt("msgcntr.no.reply.max.respond", 5);
+	  if (getNumMessageRespond(rrepMsg.getCreatedBy(), rrepMsg.getInReplyTo().getId()) > max) {
+		  return true;
+	  }
+	  return false;
+  }
+  
+  private int getNumMessageRespond(final String userId, final Long messageId) {
+	  log.debug("getNumMessageRespond executing");
+
+	  HibernateCallback<Number> hcb = session -> {
+		  Query q = session.getNamedQuery(QUERY_RESPONSED_COUNT);
+		  q.setParameter("userId", userId, StringType.INSTANCE);
+		  q.setParameter("messageId", messageId, LongType.INSTANCE);
+		  return (Number) q.uniqueResult();
+	  };
+    
+	  return getHibernateTemplate().execute(hcb).intValue();
+  }
+  
+  private String getUserIdByFowardMail(final Long messageId, final String mail) {
+	  log.debug("getUserIdByFowardMail executing");
+
+	  HibernateCallback<String> hcb = session -> {
+		  Query q = session.getNamedQuery(QUERY_USER_ID_BY_FOWARD_MAIL);
+		  q.setParameter("messageId", messageId, LongType.INSTANCE);
+		  q.setParameter("mail", mail, StringType.INSTANCE);
+		  return (String)q.uniqueResult();
+	  };
+
+	  return getHibernateTemplate().execute(hcb);
+  }
+
+  public void processPvtMsgReplySentAction(PrivateMessage currentMessage, PrivateMessage rrepMsg) {
+	  if (rrepMsg != null) {
+		  Map<User, Boolean> recipients = getRecipients(rrepMsg.getRecipients());
+
+		  sendPrivateMessage(rrepMsg, recipients, false);
+
+		  if (!rrepMsg.getDraft()) {
+			  markMessageAsRepliedForUser(currentMessage, rrepMsg.getAuthorId());
+			  String contextId = ((PrivateMessageRecipientImpl)currentMessage.getRecipients().get(0)).getContextId();
+			  markMessageAsReadForUser(currentMessage, contextId, rrepMsg.getAuthorId(), DiscussionForumService.MESSAGES_TOOL_ID);
+			  LRS_Statement statement = null;
+			  if (null != learningResourceStoreService) {
```

Comment:
should this ever be null?

Reply:
This implementation has been ported from /messageforums-tool/src/java/org/sakaiproject/tool/messageforums/PrivateMessagesTool.java, if there is a possibility that the object is null, since the call comes from mailarchive-subetha

Reply:
It will never be null so this is superfluous